internal/cni-server/cni-plugin.go

```diff
@@ -256,79 +257,81 @@ func (s *server) listenConfig(addr net.Addr, netns string) net.ListenConfig {
+		return err
+	}
+	for _, f := range hostProc {
+		if _, err = strconv.Atoi(f.Name()); err == nil {
+			pid := f.Name()
+			netns, err := ns.GetNS(fmt.Sprintf("%s/%s/ns/net", config.HostProc, pid))
+			if err != nil {
+				log.Errorf("Failed to get ns for %s", netns.Path())
+				continue
+			}
+			if skipListening(pid) {
+				// ignore uninjected pods
+				log.Debugf("skip listening for pid(%s)", pid)
+				continue
+			}
+			if err = netns.Do(func(_ ns.NetNS) error {
+				log.Infof("build listener for pid(%s)", pid)
+				// listen on 39807
+				return s.buildListener(netns.Path())
+			}); err != nil {
+				return err
+			}
+			// attach xdp to the veth pair
+			if err = netnsPairEthDo(netns.Path(), func(name string, index int) error {
+				xdp, err := ebpf.LoadPinnedProgram(path.Join(s.bpfMountPath, "mb_xdp"), &ebpf.LoadPinOptions{})
+				// todo support load by ID: xdp, err := ebpf.NewProgramFromID(1595)
+				if err != nil {
+					log.Errorf("Failed to load %s: %v", path.Join(s.bpfMountPath, "mb_xdp"), err)
+					return err
+				}
 
-// 					l, err := link.AttachXDP(link.XDPOptions{
-// 						Program:   xdp,
-// 						Interface: index,
-// 						Flags:     link.XDPGenericMode,
-// 					})
-// 					if err != nil {
-// 						return err
-// 					}
-// 					p := getXDPPinnedPath(s.bpfMountPath, f.Name(), f.Name(), name)
-// 					_ = os.MkdirAll(p, os.ModePerm)
-// 					return l.Pin(path.Join(p, "mb_xdp"))
-// 				}); err != nil {
-// 					return err
-// 				}
-// 			}
-// 		}
-// 	}
-// 	return nil
-// }
+				l, err := link.AttachXDP(link.XDPOptions{
+					Program:   xdp,
+					Interface: index,
+					Flags:     link.XDPGenericMode,
+				})
+				if err != nil {
+					log.Errorf("Failed to attach xdp to interface (index: %d): %v", index, err)
+					return err
+				}
+				p := getXDPPinnedPath(s.bpfMountPath, pid, pid, name)
+				_ = os.MkdirAll(p, os.ModePerm)
+				return l.Pin(path.Join(p, "mb_xdp"))
+			}); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
 
-// func ignore() bool {
-// 	conn1, err := net.Dial("tcp", "0.0.0.0:15001")
-// 	if err != nil {
-// 		// uninjected port
+func skipListening(pid string) bool {
+	b, _ := os.ReadFile(fmt.Sprintf("%s/%s/comm", config.HostProc, pid))
+	comm := string(b)
+	if comm != "pilot-agent\n" {
+		return true
+	}
+	b, _ = os.ReadFile(fmt.Sprintf("%s/%s/net/tcp", config.HostProc, pid))
+	conn4 := string(b)
+	// local_address = 0.0.0.0:15001
+	if !strings.Contains(conn4, ": 00000000:3A99 00000000:0000 0A") {
+		// uninjected pod
+		return true
+	}
 
-// 		// conn6, err := net.Dial("tcp", "[::]:15001")
-// 		// defer conn6.Close()
-// 		// if err != nil {
-// 		// 	return true
-// 		// }
-// 		return true
-// 	}
-// 	conn1.Close()
-// 	conn2, err := net.Dial("tcp", "0.0.0.0:39807")
-// 	if err != nil {
-// 		return false
-// 	}
-// 	// if it has listened on 39807, do nothing
-// 	conn2.Close()
-// 	return true
-// }
+	b, _ = os.ReadFile(fmt.Sprintf("%s/%s/net/tcp6", config.HostProc, pid))
```

Comment:
In most cases, we do not need to read the entire file content completely, and getting a batch read match may be a better choice.
In a real environment, it may be more expensive to read the whole file.